java/ql/src/Telemetry/AutomodelSharedCharacteristics.qll

```diff
@@ -0,0 +1,281 @@
+   * This predicate should hold for that type, and that type only.
+   */
+  predicate isNegative(EndpointType t);
+
+  /**
+   * Should hold for any endpoint that is a flow sanitizer.
+   */
+  predicate isSanitizer(Endpoint e, EndpointType t);
+
+  /**
+   * Should hold for any endpoint that is a sink of the given (known or unknown) label.
+   */
+  predicate isSink(Endpoint e, string label);
+
+  /**
+   * Should hold for any endpoint that is known to not be any sink.
+   */
+  predicate isNeutral(Endpoint e);
+
+  /**
+   * Holds if `e` has the given metadata.
+   *
+   * This is a helper function to extract and export needed information about each endpoint in the sink candidate query
+   * as well as the queries that extract positive and negative examples for the prompt / training set. The metadata is
+   * extracted as a string in the format of a Python dictionary, eg.:
+   *
+   *     `{'Package': 'com.foo.util', 'Type': 'HelperClass', ... }`.
+   *
+   * The meta data will be passed on to the machine learning code by the extraction queries.
+   */
+  predicate hasMetadata(Endpoint e, string metadata);
+}
+
+/**
+ * A set of shared characteristics for a given candidate class.
+ *
+ * This module is language-agnostic, although the `CandidateSig` module will be language-specific.
+ *
+ * The language specific implementation can also further extend the behavior of this module by adding additional
+ *   implementations of endpoint characteristics exported by this module.
+ */
+module SharedCharacteristics<CandidateSig Candidate> {
+  predicate isNegative(Candidate::EndpointType e) { Candidate::isNegative(e) }
+
+  predicate isSink(Candidate::Endpoint e, string label) { Candidate::isSink(e, label) }
+
+  predicate isNeutral(Candidate::Endpoint e) { Candidate::isNeutral(e) }
+
+  /**
+   * Holds if `sink` is a known sink of type `endpointType`.
+   */
+  predicate isKnownSink(Candidate::Endpoint sink, Candidate::EndpointType endpointType) {
+    // If the list of characteristics includes positive indicators with maximal confidence for this class, then it's a
+    // known sink for the class.
+    not isNegative(endpointType) and
+    exists(EndpointCharacteristic characteristic |
+      characteristic.appliesToEndpoint(sink) and
+      characteristic.hasImplications(endpointType, true, maximalConfidence())
+    )
+  }
+
+  /**
+   * Holds if the candidate sink `candidateSink` should be considered as a possible sink of type `sinkType`, and
+   * classified by the ML model. A candidate sink is a node that cannot be excluded from `sinkType` based on its
+   * characteristics.
+   */
+  predicate isSinkCandidate(Candidate::Endpoint candidateSink, Candidate::EndpointType sinkType) {
+    not isNegative(sinkType) and
+    not exists(getAReasonSinkExcluded(candidateSink, sinkType))
+  }
+
+  predicate hasMetadata(Candidate::Endpoint n, string metadata) {
+    Candidate::hasMetadata(n, metadata)
+  }
```

Comment:
I _think_ you can do this kind of thing with aliases, something like:
```ql
predicate hasMetadata = Candidate::hasMetadata/2;
```

Reply:
thank you, that did the trick!